### PR TITLE
witness node accepts private key from file

### PR DESCRIPTION
**Pull request by crypto-ape**
Hey Monkeys!

This change implements an option for adding private keys from files.

Ape out!

### CODE REVIEW

File: libraries/plugins/witness/witness.cpp [hunk 1]
```diff
@@ -83,35 +88,62 @@ std::string witness_plugin::plugin_name()const
...
-            {
-               private_key = fc::variant(key_id_to_wif_pair.second, 2).as<fc::ecc::private_key>(1);
-            }
-            catch (const fc::exception&)
-            {
-               FC_THROW("Invalid WIF-format private key ${key_string}", ("key_string", key_id_to_wif_pair.second));
-            }
+            elog("Failed to load private key file from ${path}",
+            ("path", key_id_to_wif_pair_file.string()));
+            std::exit(EXIT_FAILURE);
```
Comment (abitmore) [inline libraries/plugins/witness/witness.cpp]: We don't call `exit` in a library. Please raise the error normally so it can be handled by the caller.

File: libraries/plugins/witness/witness.cpp [hunk 2]
```diff
@@ -83,35 +88,62 @@ std::string witness_plugin::plugin_name()const
...
-            try
-            {
-               private_key = fc::variant(key_id_to_wif_pair.second, 2).as<fc::ecc::private_key>(1);
-            }
-            catch (const fc::exception&)
-            {
-               FC_THROW("Invalid WIF-format private key ${key_string}", ("key_string", key_id_to_wif_pair.second));
-            }
+            elog("Failed to load private key file from ${path}",
+            ("path", key_id_to_wif_pair_file.string()));
```
Comment (abitmore) [inline libraries/plugins/witness/witness.cpp]: Please indent.

File: libraries/plugins/witness/witness.cpp [hunk 3]
```diff
@@ -83,35 +88,62 @@ std::string witness_plugin::plugin_name()const
...
-               (key_id_to_wif_pair_string, 5);
-         ilog("Public Key: ${public}", ("public", key_id_to_wif_pair.first));
-         fc::optional<fc::ecc::private_key> private_key = graphene::utilities::wif_to_key(key_id_to_wif_pair.second);
-         if (!private_key)
+         add_private_key(key_id_to_wif_pair_string);
+      }
+   }
+   if (options.count("private-key-file"))
+   {
+      const std::vector<boost::filesystem::path> key_id_to_wif_pair_files = options["private-key-file"].as<std::vector<boost::filesystem::path>>();
```
Comment (abitmore) [inline libraries/plugins/witness/witness.cpp]: Please wrap this long line (ideal max line length = 118 chars).

File: libraries/plugins/witness/witness.cpp [hunk 4]
```diff
@@ -83,35 +88,62 @@ std::string witness_plugin::plugin_name()const
...
+}
+
 void witness_plugin::plugin_initialize(const boost::program_options::variables_map& options)
 { try {
    ilog("witness plugin:  plugin_initialize() begin");
    _options = &options;
    LOAD_VALUE_SET(options, "witness-id", _witnesses, chain::witness_id_type)
 
-   if( options.count("private-key") )
+   if (options.count("private-key"))
```
Comment (abitmore) [inline libraries/plugins/witness/witness.cpp]: Why this change? :P

File: libraries/plugins/witness/witness.cpp [hunk 5]
```diff
@@ -83,35 +88,62 @@ std::string witness_plugin::plugin_name()const
    return "witness";
 }
 
+void witness_plugin::add_private_key(const std::string& key_id_to_wif_pair_string)
+{
+   auto key_id_to_wif_pair = graphene::app::dejsonify<std::pair<chain::public_key_type, std::string> >(key_id_to_wif_pair_string, 5);
```
Comment (abitmore) [inline libraries/plugins/witness/witness.cpp]: please wrap this long line.

**Comment by pmconrad** [comment]
Any insight why the key must be specified as a [public,private] pair? The public part is redundant, it would be simpler if the file contained only private keys.

While you're at it, perhaps add another option that reads private keys from a file descriptor, as in https://www.gnupg.org/gph/en/manual/r1736.html . That would open the possibility to enter the key on startup, so that it doesn't have to be stored on disk.

File: libraries/plugins/witness/witness.cpp [hunk 6]
```diff
@@ -74,6 +77,8 @@ void witness_plugin::plugin_set_program_options(
           DEFAULT_VALUE_VECTOR(std::make_pair(chain::public_key_type(default_priv_key.get_public_key()),
                 graphene::utilities::key_to_wif(default_priv_key))),
                 "Tuple of [PublicKey, WIF private key] (may specify multiple times)")
+         ("private-key-file", bpo::value<vector<boost::filesystem::path>>()->composing()->multitoken(),
+          "Path to file containing tuples of [PublicKey, WIF private key] (may specify multiple times)")
```
Comment (pmconrad) [inline libraries/plugins/witness/witness.cpp]: AFAICS the implementation currently allows only one such pair per file. Or did I miss something?

**Comment by crypto-ape** [comment]
I guess most of the comments shall be resolved. Furthermore, I have rebased the changes on develop.

File: libraries/plugins/witness/witness.cpp (hunk 4 above)
Comment (crypto-ape) [reply libraries/plugins/witness/witness.cpp]: Well, I consider the `if(` style to be really bad. It was close to other changes... Reverted ;).

File: libraries/plugins/witness/witness.cpp (hunk 6 above)
Comment (crypto-ape) [reply libraries/plugins/witness/witness.cpp]: Yes, multiple files, one key pair per file.

Comment (pmconrad) [reply libraries/plugins/witness/witness.cpp]: Please either allow multiple pairs, or replace "tuples" above. (IMO allowing only one pair per file is counterintuitive, so the explanation text should stress that limitation.)

Comment (crypto-ape) [reply libraries/plugins/witness/witness.cpp]: A very good objection. I have modified the code to read one pair per line. Please, inspect it.

Comment (pmconrad) [reply libraries/plugins/witness/witness.cpp]: Thanks. Please modify the explanation so the user understands that each line must contain exactly one such pair.

Comment (crypto-ape) [reply libraries/plugins/witness/witness.cpp]: Done.

// Welcome back from holiday, I hope. :)

File: libraries/plugins/witness/witness.cpp [hunk 7]
```diff
@@ -74,6 +77,8 @@ void witness_plugin::plugin_set_program_options(
           DEFAULT_VALUE_VECTOR(std::make_pair(chain::public_key_type(default_priv_key.get_public_key()),
                 graphene::utilities::key_to_wif(default_priv_key))),
                 "Tuple of [PublicKey, WIF private key] (may specify multiple times)")
+         ("private-key-file", bpo::value<vector<boost::filesystem::path>>()->composing()->multitoken(),
+          "Path to file containing tuples of [PublicKey, WIF private key] (exactly one per line)")
```
Comment (abitmore) [inline libraries/plugins/witness/witness.cpp]: So, we can have multiple files, and in each file we can have multiple keys? I think we need to describe it in the string.

File: libraries/plugins/witness/witness.cpp (hunk 6 above)
Comment (pmconrad) [reply libraries/plugins/witness/witness.cpp]: Thanks. :)

File: libraries/plugins/witness/witness.cpp (hunk 7 above)
Comment (crypto-ape) [reply libraries/plugins/witness/witness.cpp]:
```cpp
("private-key-file", bpo::value<vector<boost::filesystem::path>>()->composing()->multitoken(),
"Path to a file containing tuples of [PublicKey, WIF private key]."
" The file has to contain exactly one tuple (i.e. private - public key pair) per line."
" This option may be specified multiple times, thus multiple files can be provided.")
```
Will this make it?

Comment (abitmore) [reply libraries/plugins/witness/witness.cpp]: Yes, thanks.

Comment (crypto-ape) [reply libraries/plugins/witness/witness.cpp]: Done.

Comment (abitmore) [reply libraries/plugins/witness/witness.cpp]: Thanks.
